File: node/src/config/shared.js
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const _cache = { config: null };

export function loadSharedConfig() {
  if (_cache.config) return _cache.config;

  let dir = path.dirname(fileURLToPath(import.meta.url));
  for (let i = 0; i < 10; i++) {
    const candidate = path.join(dir, "wrivon.config.json");
    if (fs.existsSync(candidate)) {
      _cache.config = JSON.parse(fs.readFileSync(candidate, "utf8"));
      return _cache.config;
    }
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }

  // Fallback: next to node/ directory
  const nodeRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..", "..", "..");
  const repoRoot = path.resolve(nodeRoot, "..");
  const fallback = path.join(repoRoot, "wrivon.config.json");
  if (fs.existsSync(fallback)) {
    _cache.config = JSON.parse(fs.readFileSync(fallback, "utf8"));
    return _cache.config;
  }

  throw new Error(
    "wrivon.config.json not found. Expected it at the repository root next to node/ and python/ directories."
  );
}

export function getProviderOrder() {
  return loadSharedConfig().provider_fallback_order || [];
}

export function getCommands() {
  return loadSharedConfig().commands || {};
}

export function getChatModes() {
  return loadSharedConfig().chat_modes || {};
}

export function getTools() {
  return loadSharedConfig().tools || {};
}

export function getDefaults() {
  return loadSharedConfig().defaults || {};
}

File: python/wrivon/shared.py
"""Shared config loader — reads wrivon.config.json from the repo root."""

import json
from pathlib import Path

_cache = {"config": None}


def load_shared_config():
    """Find and load wrivon.config.json by walking up directory tree."""
    if _cache["config"] is not None:
        return _cache["config"]

    current = Path(__file__).resolve().parent  # python/wrivon/
    for _ in range(10):
        candidate = current / "wrivon.config.json"
        if candidate.exists():
            _cache["config"] = json.loads(candidate.read_text(encoding="utf-8"))
            return _cache["config"]
        parent = current.parent
        if parent == current:
            break
        current = parent

    # Fallback: check next to repo root markers
    current = Path(__file__).resolve().parent.parent.parent  # repo root
    candidate = current / "wrivon.config.json"
    if candidate.exists():
        _cache["config"] = json.loads(candidate.read_text(encoding="utf-8"))
        return _cache["config"]

    raise FileNotFoundError(
        "wrivon.config.json not found. Expected it at the repository root "
        "next to node/ and python/ directories."
    )


def get_provider_order():
    return load_shared_config().get("provider_fallback_order", [])


def get_commands():
    return load_shared_config().get("commands", {})


def get_chat_modes():
    return load_shared_config().get("chat_modes", {})


def get_tools():
    return load_shared_config().get("tools", {})


def get_defaults():
    return load_shared_config().get("defaults", {})
